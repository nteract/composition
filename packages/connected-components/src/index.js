// @flow

export {
  default as NewNotebookNavigation,
  PureNewNotebookNavigation
} from "./new-notebook-navigation";

export { default as NotebookMenu } from "./notebook-menu";
export { default as ModalController } from "./modal-controller";
