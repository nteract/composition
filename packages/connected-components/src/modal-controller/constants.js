// @flow
export const MODAL_TYPES = {
  ABOUT: "core/about-modal"
};
