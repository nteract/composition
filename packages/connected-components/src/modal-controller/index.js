// @flow
import * as React from "react";
import { MODAL_TYPES } from "./constants";
import AboutModal from "./about-modal";
import { connect } from "react-redux";
import { selectors } from "@nteract/core";

class ModalController extends React.Component<*, *> {
  getModal = () => {
    const { modalType } = this.props;
    switch (modalType) {
      case MODAL_TYPES.ABOUT:
        return AboutModal;
      default:
        return null;
    }
  };
  render() {
    const Modal = this.getModal();
    return Modal ? <Modal /> : null;
  }
}

import type { MapStateToProps } from "react-redux";

const mapStateToProps: MapStateToProps<*, *, *> = (state: *) => ({
  modalType: selectors.modalType(state)
});

export { MODAL_TYPES };

export default connect(mapStateToProps)(ModalController);
