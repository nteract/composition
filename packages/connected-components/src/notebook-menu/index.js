// @flow
import * as Immutable from "immutable";
import * as React from "react";
import Menu, { SubMenu, Divider, MenuItem } from "rc-menu";
import { actions, selectors } from "@nteract/core";
import type {
  AppState,
  ContentRef,
  KernelRef,
  KernelspecsRef,
  KernelspecsByRefRecord
} from "@nteract/core";
import { connect } from "react-redux";

import { MODAL_TYPES } from "../modal-controller";

import { MENU_ITEM_ACTIONS, MENUS } from "./constants";
import StyleWrapper from "./styles";

// To allow actions that can take dynamic arguments (like selecting a kernel
// based on the host's kernelspecs), we have some simple utility functions to
// stringify/parse actions/arguments.
const createActionKey = (action, ...args) => [action, ...args].join(":");
const parseActionKey = key => key.split(":");

type Props = {
  persistAfterClick?: boolean,
  defaultOpenKeys?: Array<string>,
  openKeys?: Array<string>,
  currentKernelRef: ?KernelRef,
  saveNotebook: ?(payload: *) => void,
  downloadNotebook: ?(payload: *) => void,
  executeCell: ?(payload: *) => void,
  executeAllCells: ?(payload: *) => void,
  executeAllCellsBelow: ?(payload: *) => void,
  clearAllOutputs: ?(payload: *) => void,
  unhideAll: ?(*) => void,
  cutCell: ?(payload: *) => void,
  copyCell: ?(payload: *) => void,
  notebook: Immutable.Map<string, *>,
  pasteCell: ?(payload: *) => void,
  createCellBelow: ?(payload: *) => void,
  changeCellType: ?(payload: *) => void,
  setTheme: ?(theme: ?string) => void,
  openAboutModal: ?() => void,
  changeKernelByName: ?(payload: {
    kernelSpecName: string,
    oldKernelRef: ?KernelRef,
    contentRef: ContentRef
  }) => void,
  restartKernel: ?(payload: *) => void,
  restartKernelAndClearOutputs: ?(payload: *) => void,
  restartKernelAndRunAllOutputs: ?(payload: *) => void,
  killKernel: ?(payload: *) => void,
  interruptKernel: ?(payload: *) => void,
  currentContentRef: ContentRef,
  currentKernelspecsRef: ?KernelspecsRef,
  currentKernelspecs: ?KernelspecsByRefRecord
};

type State = {
  openKeys?: Array<string>
};

class PureNotebookMenu extends React.Component<Props, State> {
  static defaultProps = {
    cellFocused: null,
    saveNotebook: null,
    downloadNotebook: null,
    currentKernelRef: null,
    executeCell: null,
    executeAllCells: null,
    executeAllCellsBelow: null,
    clearAllOutputs: null,
    unhideAll: null,
    cutCell: null,
    copyCell: null,
    pasteCell: null,
    createCellBelow: null,
    setCellTypeCode: null,
    setCellTypeMarkdown: null,
    setTheme: null,
    openAboutModal: null,
    changeKernelByName: null,
    restartKernel: null,
    killKernel: null,
    interruptKernel: null,
    currentKernelspecsRef: null,
    currentKernelspecs: null
  };
  state = {};
  handleClick = ({ key }: { key: string }) => {
    const {
      persistAfterClick,
      saveNotebook,
      downloadNotebook,
      changeKernelByName,
      currentKernelRef,
      copyCell,
      createCellBelow,
      cutCell,
      executeAllCells,
      executeAllCellsBelow,
      clearAllOutputs,
      unhideAll,
      openAboutModal,
      pasteCell,
      setTheme,
      changeCellType,
      restartKernel,
      restartKernelAndClearOutputs,
      restartKernelAndRunAllOutputs,
      killKernel,
      interruptKernel,
      currentContentRef
    } = this.props;
    const [action, ...args] = parseActionKey(key);
    switch (action) {
      case MENU_ITEM_ACTIONS.SAVE_NOTEBOOK:
        if (saveNotebook) {
          saveNotebook({ contentRef: currentContentRef });
        }
        break;
      case MENU_ITEM_ACTIONS.DOWNLOAD_NOTEBOOK:
        if (downloadNotebook) {
          downloadNotebook({ contentRef: currentContentRef });
        }
        break;
      case MENU_ITEM_ACTIONS.COPY_CELL:
        if (copyCell) {
          copyCell({ contentRef: currentContentRef });
        }
        break;
      case MENU_ITEM_ACTIONS.CUT_CELL:
        if (cutCell) {
          cutCell({ contentRef: currentContentRef });
        }
        break;
      case MENU_ITEM_ACTIONS.PASTE_CELL:
        if (pasteCell) {
          pasteCell({ contentRef: currentContentRef });
        }
        break;
      case MENU_ITEM_ACTIONS.CREATE_CODE_CELL:
        if (createCellBelow) {
          createCellBelow({
            cellType: "code",
            source: "",
            contentRef: currentContentRef
          });
        }
        break;
      case MENU_ITEM_ACTIONS.CREATE_MARKDOWN_CELL:
        if (createCellBelow) {
          createCellBelow({
            cellType: "markdown",
            source: "",
            contentRef: currentContentRef
          });
        }
        break;
      case MENU_ITEM_ACTIONS.SET_CELL_TYPE_CODE:
        if (changeCellType) {
          changeCellType({
            to: "code",
            contentRef: currentContentRef
          });
        }
        break;
      case MENU_ITEM_ACTIONS.SET_CELL_TYPE_MARKDOWN:
        if (changeCellType) {
          changeCellType({
            to: "markdown",
            contentRef: currentContentRef
          });
        }
        break;
      case MENU_ITEM_ACTIONS.EXECUTE_ALL_CELLS:
        if (executeAllCells) {
          executeAllCells({ contentRef: currentContentRef });
        }
        break;
      case MENU_ITEM_ACTIONS.EXECUTE_ALL_CELLS_BELOW:
        if (executeAllCellsBelow) {
          executeAllCellsBelow({ contentRef: currentContentRef });
        }
        break;
      case MENU_ITEM_ACTIONS.UNHIDE_ALL:
        if (unhideAll) {
          unhideAll({
            outputHidden: false,
            inputHidden: false,
            contentRef: currentContentRef
          });
        }
        break;
      case MENU_ITEM_ACTIONS.CLEAR_ALL_OUTPUTS:
        if (clearAllOutputs) {
          clearAllOutputs({ contentRef: currentContentRef });
        }
        break;
      case MENU_ITEM_ACTIONS.SET_THEME_DARK:
        if (setTheme) {
          setTheme("dark");
        }
        break;
      case MENU_ITEM_ACTIONS.SET_THEME_LIGHT:
        if (setTheme) {
          setTheme("light");
        }
        break;
      case MENU_ITEM_ACTIONS.OPEN_ABOUT:
        if (openAboutModal) {
          openAboutModal();
        }
        break;
      case MENU_ITEM_ACTIONS.INTERRUPT_KERNEL:
        if (interruptKernel) {
          interruptKernel({ kernelRef: currentKernelRef });
        }
        break;
      case MENU_ITEM_ACTIONS.RESTART_KERNEL:
        if (restartKernel) {
          restartKernel({
            kernelRef: currentKernelRef,
            contentRef: currentContentRef
          });
        }
        break;
      case MENU_ITEM_ACTIONS.RESTART_AND_CLEAR_OUTPUTS:
        if (restartKernelAndClearOutputs) {
          restartKernelAndClearOutputs({
            kernelRef: currentKernelRef,
            contentRef: currentContentRef
          });
        }
        break;
      case MENU_ITEM_ACTIONS.RESTART_AND_RUN_ALL_OUTPUTS:
        if (restartKernelAndRunAllOutputs) {
          restartKernelAndRunAllOutputs({
            kernelRef: currentKernelRef,
            contentRef: currentContentRef
          });
        }
        break;
      case MENU_ITEM_ACTIONS.KILL_KERNEL:
        if (killKernel) {
          killKernel({ kernelRef: currentKernelRef });
        }
        break;
      case MENU_ITEM_ACTIONS.CHANGE_KERNEL:
        if (changeKernelByName) {
          changeKernelByName({
            oldKernelRef: currentKernelRef,
            contentRef: currentContentRef,
            kernelSpecName: args[0]
          });
        }
        break;
      default:
        console.log(`unhandled action: ${action}`);
    }

    if (!persistAfterClick) {
      this.setState({ openKeys: [] });
    }
  };
  handleOpenChange = (openKeys: Array<string>) => {
    if (!this.props.persistAfterClick) {
      this.setState({ openKeys });
    }
  };
  componentWillMount() {
    // This ensures that we can still initially set defaultOpenKeys when
    // persistAfterClick is true.
    this.setState({ openKeys: this.props.defaultOpenKeys });
  }
  render() {
    const {
      currentKernelspecs,
      defaultOpenKeys,
      persistAfterClick
    } = this.props;
    const { openKeys } = this.state;
    const menuProps: Object = {
      mode: "horizontal",
      onClick: this.handleClick,
      onOpenChange: this.handleOpenChange,
      defaultOpenKeys: defaultOpenKeys,
      selectable: false
    };
    if (!persistAfterClick) {
      menuProps.openKeys = openKeys;
    }
    return (
      <StyleWrapper>
        <Menu {...menuProps}>
          <SubMenu key={MENUS.FILE} title="File">
            <MenuItem>
              <a
                href="/nteract/edit"
                style={{
                  textDecoration: "none",
                  color: "currentColor"
                }}
                target="_blank"
              >
                Open...
              </a>
            </MenuItem>
            <MenuItem key={createActionKey(MENU_ITEM_ACTIONS.SAVE_NOTEBOOK)}>
              Save
            </MenuItem>
            <MenuItem
              key={createActionKey(MENU_ITEM_ACTIONS.DOWNLOAD_NOTEBOOK)}
            >
              Download (.ipynb)
            </MenuItem>
          </SubMenu>
          <SubMenu key={MENUS.EDIT} title="Edit">
            <MenuItem key={createActionKey(MENU_ITEM_ACTIONS.CUT_CELL)}>
              Cut Cell
            </MenuItem>
            <MenuItem key={createActionKey(MENU_ITEM_ACTIONS.COPY_CELL)}>
              Copy Cell
            </MenuItem>
            <MenuItem key={createActionKey(MENU_ITEM_ACTIONS.PASTE_CELL)}>
              Paste Cell Below
            </MenuItem>
            <Divider />
            <SubMenu key={MENUS.EDIT_SET_CELL_TYPE} title="Cell Type">
              <MenuItem
                key={createActionKey(MENU_ITEM_ACTIONS.SET_CELL_TYPE_CODE)}
              >
                Code
              </MenuItem>
              <MenuItem
                key={createActionKey(MENU_ITEM_ACTIONS.SET_CELL_TYPE_MARKDOWN)}
              >
                Markdown
              </MenuItem>
            </SubMenu>
          </SubMenu>
          <SubMenu key={MENUS.VIEW} title="View">
            <SubMenu key={MENUS.VIEW_THEMES} title="themes">
              <MenuItem
                key={createActionKey(MENU_ITEM_ACTIONS.SET_THEME_LIGHT)}
              >
                light
              </MenuItem>
              <MenuItem key={createActionKey(MENU_ITEM_ACTIONS.SET_THEME_DARK)}>
                dark
              </MenuItem>
            </SubMenu>
          </SubMenu>
          <SubMenu key={MENUS.CELL} title="Cell">
            <MenuItem
              key={createActionKey(MENU_ITEM_ACTIONS.EXECUTE_ALL_CELLS)}
            >
              Run All Cells
            </MenuItem>
            <MenuItem
              key={createActionKey(MENU_ITEM_ACTIONS.EXECUTE_ALL_CELLS_BELOW)}
            >
              Run All Cells Below
            </MenuItem>
            <Divider />
            <SubMenu key={MENUS.CELL_CREATE_CELL} title="New Cell">
              <MenuItem
                key={createActionKey(MENU_ITEM_ACTIONS.CREATE_CODE_CELL)}
              >
                Code
              </MenuItem>
              <MenuItem
                key={createActionKey(MENU_ITEM_ACTIONS.CREATE_MARKDOWN_CELL)}
              >
                Markdown
              </MenuItem>
            </SubMenu>
            <Divider />

            <MenuItem
              key={createActionKey(MENU_ITEM_ACTIONS.CLEAR_ALL_OUTPUTS)}
            >
              Clear All Outputs
            </MenuItem>

            <MenuItem key={createActionKey(MENU_ITEM_ACTIONS.UNHIDE_ALL)}>
              Unhide All Input and Output
            </MenuItem>
          </SubMenu>

          <SubMenu key={MENUS.RUNTIME} title="Runtime">
            <MenuItem key={createActionKey(MENU_ITEM_ACTIONS.INTERRUPT_KERNEL)}>
              Interrupt
            </MenuItem>
            <Divider />
            <MenuItem key={createActionKey(MENU_ITEM_ACTIONS.KILL_KERNEL)}>
              Halt
            </MenuItem>
            <MenuItem key={createActionKey(MENU_ITEM_ACTIONS.RESTART_KERNEL)}>
              Restart
            </MenuItem>
            <MenuItem
              key={createActionKey(MENU_ITEM_ACTIONS.RESTART_AND_CLEAR_OUTPUTS)}
            >
              Restart and Clear All Cells
            </MenuItem>
            <MenuItem
              key={createActionKey(
                MENU_ITEM_ACTIONS.RESTART_AND_RUN_ALL_OUTPUTS
              )}
            >
              Restart and Run All Cells
            </MenuItem>
            <Divider />
            <SubMenu
              key={MENUS.RUNTIME_CHANGE_KERNEL}
              title="Change Kernel"
              disabled={!currentKernelspecs}
            >
              {currentKernelspecs
                ? currentKernelspecs.byName
                    .keySeq()
                    .map(name => [
                      name,
                      currentKernelspecs.byName.getIn([name, "displayName"])
                    ])
                    .toArray()
                    .map(([name, displayName]) => {
                      return (
                        <MenuItem
                          key={createActionKey(
                            MENU_ITEM_ACTIONS.CHANGE_KERNEL,
                            name
                          )}
                        >
                          {displayName}
                        </MenuItem>
                      );
                    })
                : null}
            </SubMenu>
          </SubMenu>

          <SubMenu key={MENUS.HELP} title="Help">
            <MenuItem key={createActionKey(MENU_ITEM_ACTIONS.OPEN_ABOUT)}>
              About
            </MenuItem>
          </SubMenu>
        </Menu>
        <style jsx>{`
          position: sticky;
          top: 0;
          /* TODO: this is getting ridiculous... */
          z-index: 10000;
        `}</style>
      </StyleWrapper>
    );
  }
}

const mapStateToProps = (
  state: AppState,
  { contentRef }: { contentRef: ContentRef }
) => {
  return {
    currentKernelRef: selectors.currentKernelRef(state),
    currentContentRef: contentRef,
    currentKernelspecsRef: selectors.currentKernelspecsRef(state),
    currentKernelspecs: selectors.currentKernelspecs(state)
  };
};

const mapDispatchToProps = dispatch => ({
  saveNotebook: (payload: *) => dispatch(actions.save(payload)),
  downloadNotebook: (payload: *) => dispatch(actions.downloadContent(payload)),
  executeCell: (payload: *) => dispatch(actions.executeCell(payload)),
  executeAllCells: (payload: *) => dispatch(actions.executeAllCells(payload)),
  executeAllCellsBelow: (payload: *) =>
    dispatch(actions.executeAllCellsBelow(payload)),
  clearAllOutputs: (payload: *) => dispatch(actions.clearAllOutputs(payload)),
  unhideAll: payload => dispatch(actions.unhideAll(payload)),
  cutCell: (payload: *) => dispatch(actions.cutCell(payload)),
  copyCell: (payload: *) => dispatch(actions.copyCell(payload)),
  pasteCell: (payload: *) => dispatch(actions.pasteCell(payload)),
  createCellBelow: (payload: *) => dispatch(actions.createCellBelow(payload)),
  changeCellType: (payload: *) => dispatch(actions.changeCellType(payload)),
  setTheme: theme => dispatch(actions.setTheme(theme)),
  openAboutModal: () =>
    dispatch(actions.openModal({ modalType: MODAL_TYPES.ABOUT })),
  changeKernelByName: payload => dispatch(actions.changeKernelByName(payload)),
  restartKernel: payload => dispatch(actions.restartKernel(payload)),
  restartKernelAndClearOutputs: payload =>
    dispatch(
      actions.restartKernel({ ...payload, outputHandling: "Clear All" })
    ),
  restartKernelAndRunAllOutputs: payload =>
    dispatch(actions.restartKernel({ ...payload, outputHandling: "Run All" })),
  killKernel: payload => dispatch(actions.killKernel(payload)),
  interruptKernel: payload => dispatch(actions.interruptKernel(payload))
});

// $FlowFixMe: react-redux
const NotebookMenu = connect(
  mapStateToProps,
  mapDispatchToProps
)(PureNotebookMenu);

// We export this for testing purposes.
export { PureNotebookMenu };

export default NotebookMenu;
