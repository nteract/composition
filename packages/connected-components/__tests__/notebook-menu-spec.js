// @flow
import React from "react";
import renderer from "react-test-renderer";
import { shallow, mount } from "enzyme";

import { PureNotebookMenu } from "../src/notebook-menu";
import { MENU_ITEM_ACTIONS, MENUS } from "../src/notebook-menu/constants";

describe("PureNotebookMenu ", () => {
  describe("snapshots", () => {
    test("renders the default", () => {
      const component = renderer.create(<PureNotebookMenu />);
      let tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
  describe("shallow", () => {
    test("renders the default", () => {
      const wrapper = shallow(<PureNotebookMenu />);
      expect(wrapper).not.toBeNull();
    });
  });
  describe("mount", () => {
    test("renders the default", () => {
      const wrapper = mount(<PureNotebookMenu />);
      expect(wrapper).not.toBeNull();
    });
    test("calls appropriate handlers on click", () => {
      const props = {
        // keep menu open after clicks
        persistAfterClick: true,

        // action functions
        executeCell: jest.fn(),
        executeAllCells: jest.fn(),
        executeAllCellsBelow: jest.fn(),
        unhideAll: jest.fn(),
        clearAllOutputs: jest.fn(),
        cutCell: jest.fn(),
        copyCell: jest.fn(),
        pasteCell: jest.fn(),
        createCellBelow: jest.fn(),
        changeCellType: jest.fn(),
        setTheme: jest.fn(),
        saveNotebook: jest.fn(),
        openAboutModal: jest.fn(),
        interruptKernel: jest.fn(),
        restartKernel: jest.fn(),
        restartKernelAndClearOutputs: jest.fn(),
        restartKernelAndRunAllOutputs: jest.fn(),
        killKernel: jest.fn(),
        downloadNotebook: jest.fn(),

        // document state (we mock out the implementation, so these are just
        // dummy variables.
        currentContentRef: "fake-content-ref",
        currentKernelRef: "fake-kernel-ref",

        // menu props, note that we force all menus to be open to click.
        defaultOpenKeys: Object.values(MENUS)
      };
      const wrapper = mount(<PureNotebookMenu {...props} />);

      const downloadNotebookItem = wrapper
        .find({ eventKey: MENU_ITEM_ACTIONS.DOWNLOAD_NOTEBOOK })
        .first();
      expect(props.downloadNotebook).not.toHaveBeenCalled();
      downloadNotebookItem.simulate("click");
      expect(props.downloadNotebook).toHaveBeenCalledTimes(1);
      expect(props.downloadNotebook).toHaveBeenCalledWith({
        contentRef: props.currentContentRef
      });

      const executeAllCellsItem = wrapper
        .find({ eventKey: MENU_ITEM_ACTIONS.EXECUTE_ALL_CELLS })
        .first();
      expect(props.executeAllCells).not.toHaveBeenCalled();
      executeAllCellsItem.simulate("click");
      expect(props.executeAllCells).toHaveBeenCalledTimes(1);
      expect(props.executeAllCells).toHaveBeenCalledWith({
        contentRef: props.currentContentRef
      });

      const executeAllCellsBelowItem = wrapper
        .find({ eventKey: MENU_ITEM_ACTIONS.EXECUTE_ALL_CELLS_BELOW })
        .first();
      expect(props.executeAllCellsBelow).not.toHaveBeenCalled();
      executeAllCellsBelowItem.simulate("click");
      expect(props.executeAllCellsBelow).toHaveBeenCalledTimes(1);
      expect(props.executeAllCellsBelow).toHaveBeenCalledWith({
        contentRef: props.currentContentRef
      });

      const unhideAllItem = wrapper
        .find({ eventKey: MENU_ITEM_ACTIONS.UNHIDE_ALL })
        .first();
      expect(props.unhideAll).not.toHaveBeenCalled();
      unhideAllItem.simulate("click");
      expect(props.unhideAll).toHaveBeenCalledTimes(1);
      expect(props.unhideAll).toHaveBeenCalledWith({
        outputHidden: false,
        inputHidden: false,
        contentRef: props.currentContentRef
      });

      const clearAllOutputsItem = wrapper
        .find({ eventKey: MENU_ITEM_ACTIONS.CLEAR_ALL_OUTPUTS })
        .first();
      expect(props.clearAllOutputs).not.toHaveBeenCalled();
      clearAllOutputsItem.simulate("click");
      expect(props.clearAllOutputs).toHaveBeenCalledTimes(1);
      expect(props.clearAllOutputs).toHaveBeenCalledWith({
        contentRef: props.currentContentRef
      });

      const cutCellItem = wrapper
        .find({ eventKey: MENU_ITEM_ACTIONS.CUT_CELL })
        .first();
      expect(props.cutCell).not.toHaveBeenCalled();
      cutCellItem.simulate("click");
      expect(props.cutCell).toHaveBeenCalledTimes(1);
      expect(props.cutCell).toHaveBeenCalledWith({
        contentRef: props.currentContentRef
      });

      const copyCellItem = wrapper
        .find({ eventKey: MENU_ITEM_ACTIONS.COPY_CELL })
        .first();
      expect(props.copyCell).not.toHaveBeenCalled();
      copyCellItem.simulate("click");
      expect(props.copyCell).toHaveBeenCalledTimes(1);
      expect(props.copyCell).toHaveBeenCalledWith({
        contentRef: props.currentContentRef
      });

      const pasteCellItem = wrapper
        .find({ eventKey: MENU_ITEM_ACTIONS.PASTE_CELL })
        .first();
      expect(props.pasteCell).not.toHaveBeenCalled();
      pasteCellItem.simulate("click");
      expect(props.pasteCell).toHaveBeenCalledTimes(1);
      expect(props.pasteCell).toHaveBeenCalledWith({
        contentRef: props.currentContentRef
      });

      const createMarkdownCellItem = wrapper
        .find({ eventKey: MENU_ITEM_ACTIONS.CREATE_MARKDOWN_CELL })
        .first();
      expect(props.createCellBelow).not.toHaveBeenCalled();
      createMarkdownCellItem.simulate("click");
      expect(props.createCellBelow).toHaveBeenCalledTimes(1);
      expect(props.createCellBelow).toHaveBeenCalledWith({
        cellType: "markdown",
        contentRef: props.currentContentRef,
        source: ""
      });

      props.createCellBelow.mockClear();
      const createCodeCellItem = wrapper
        .find({ eventKey: MENU_ITEM_ACTIONS.CREATE_CODE_CELL })
        .first();
      expect(props.createCellBelow).not.toHaveBeenCalled();
      createCodeCellItem.simulate("click");
      expect(props.createCellBelow).toHaveBeenCalledTimes(1);
      expect(props.createCellBelow).toHaveBeenCalledWith({
        cellType: "code",
        contentRef: props.currentContentRef,
        source: ""
      });

      const setCellTypeCodeItem = wrapper
        .find({ eventKey: MENU_ITEM_ACTIONS.SET_CELL_TYPE_CODE })
        .first();
      expect(props.changeCellType).not.toHaveBeenCalled();
      setCellTypeCodeItem.simulate("click");
      expect(props.changeCellType).toHaveBeenCalledTimes(1);
      expect(props.changeCellType).toHaveBeenCalledWith({
        to: "code",
        contentRef: props.currentContentRef
      });

      props.changeCellType.mockClear();
      const setCellTypeMarkdownItem = wrapper
        .find({ eventKey: MENU_ITEM_ACTIONS.SET_CELL_TYPE_MARKDOWN })
        .first();
      expect(props.changeCellType).not.toHaveBeenCalled();
      setCellTypeMarkdownItem.simulate("click");
      expect(props.changeCellType).toHaveBeenCalledTimes(1);
      expect(props.changeCellType).toHaveBeenCalledWith({
        to: "markdown",
        contentRef: props.currentContentRef
      });

      const setThemeLightItem = wrapper
        .find({ eventKey: MENU_ITEM_ACTIONS.SET_THEME_LIGHT })
        .first();
      expect(props.setTheme).not.toHaveBeenCalled();
      setThemeLightItem.simulate("click");
      expect(props.setTheme).toHaveBeenCalledTimes(1);
      expect(props.setTheme).toHaveBeenCalledWith("light");

      props.setTheme.mockClear();
      const setThemeDarkItem = wrapper
        .find({ eventKey: MENU_ITEM_ACTIONS.SET_THEME_DARK })
        .first();
      expect(props.setTheme).not.toHaveBeenCalled();
      setThemeDarkItem.simulate("click");
      expect(props.setTheme).toHaveBeenCalledTimes(1);
      expect(props.setTheme).toHaveBeenCalledWith("dark");

      const saveNotebookItem = wrapper
        .find({ eventKey: MENU_ITEM_ACTIONS.SAVE_NOTEBOOK })
        .first();
      expect(props.saveNotebook).not.toHaveBeenCalled();
      saveNotebookItem.simulate("click");
      expect(props.saveNotebook).toHaveBeenCalledTimes(1);
      expect(props.saveNotebook).toHaveBeenCalledWith({
        contentRef: props.currentContentRef
      });

      const openAboutItem = wrapper
        .find({ eventKey: MENU_ITEM_ACTIONS.OPEN_ABOUT })
        .first();
      expect(props.openAboutModal).not.toHaveBeenCalled();
      openAboutItem.simulate("click");
      expect(props.openAboutModal).toHaveBeenCalledTimes(1);
      expect(props.openAboutModal).toHaveBeenCalledWith();

      const interruptKernelItem = wrapper
        .find({ eventKey: MENU_ITEM_ACTIONS.INTERRUPT_KERNEL })
        .first();
      expect(props.interruptKernel).not.toHaveBeenCalled();
      interruptKernelItem.simulate("click");
      expect(props.interruptKernel).toHaveBeenCalledTimes(1);
      expect(props.interruptKernel).toHaveBeenCalledWith({
        kernelRef: props.currentKernelRef
      });

      const restartKernelItem = wrapper
        .find({ eventKey: MENU_ITEM_ACTIONS.RESTART_KERNEL })
        .first();
      expect(props.restartKernel).not.toHaveBeenCalled();
      restartKernelItem.simulate("click");
      expect(props.restartKernel).toHaveBeenCalledTimes(1);
      expect(props.restartKernel).toHaveBeenCalledWith({
        contentRef: props.currentContentRef,
        kernelRef: props.currentKernelRef
      });

      const restartKernelAndClearOutputsItem = wrapper
        .find({ eventKey: MENU_ITEM_ACTIONS.RESTART_AND_CLEAR_OUTPUTS })
        .first();
      expect(props.restartKernelAndClearOutputs).not.toHaveBeenCalled();
      restartKernelAndClearOutputsItem.simulate("click");
      expect(props.restartKernelAndClearOutputs).toHaveBeenCalledTimes(1);
      expect(props.restartKernelAndClearOutputs).toHaveBeenCalledWith({
        contentRef: props.currentContentRef,
        kernelRef: props.currentKernelRef
      });

      const restartKernelAndRunAllOutputsItem = wrapper
        .find({ eventKey: MENU_ITEM_ACTIONS.RESTART_AND_RUN_ALL_OUTPUTS })
        .first();
      expect(props.restartKernelAndRunAllOutputs).not.toHaveBeenCalled();
      restartKernelAndRunAllOutputsItem.simulate("click");
      expect(props.restartKernelAndRunAllOutputs).toHaveBeenCalledTimes(1);
      expect(props.restartKernelAndRunAllOutputs).toHaveBeenCalledWith({
        contentRef: props.currentContentRef,
        kernelRef: props.currentKernelRef
      });

      const killKernelItem = wrapper
        .find({ eventKey: MENU_ITEM_ACTIONS.KILL_KERNEL })
        .first();
      expect(props.killKernel).not.toHaveBeenCalled();
      killKernelItem.simulate("click");
      expect(props.killKernel).toHaveBeenCalledTimes(1);
      expect(props.killKernel).toHaveBeenCalledWith({
        kernelRef: props.currentKernelRef
      });
    });
  });
});
