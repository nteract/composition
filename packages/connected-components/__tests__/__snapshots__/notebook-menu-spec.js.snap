// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PureNotebookMenu  snapshots renders the default 1`] = `
<ul
  aria-activedescendant=""
  className="rc-menu  rc-menu-root rc-menu-horizontal"
  onKeyDown={[Function]}
  role="menu"
  style={Object {}}
  tabIndex="0"
>
  <li
    className="rc-menu-submenu rc-menu-submenu-horizontal"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    style={undefined}
  >
    <div
      aria-expanded={false}
      aria-haspopup="true"
      aria-owns="file$Menu"
      className="rc-menu-submenu-title"
      onBlur={undefined}
      onClick={[Function]}
      onContextMenu={undefined}
      onFocus={undefined}
      onMouseDown={undefined}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onTouchStart={undefined}
      style={Object {}}
      title="File"
    >
      File
      <i
        className="rc-menu-submenu-arrow"
      />
    </div>
  </li>
  <li
    className="rc-menu-submenu rc-menu-submenu-horizontal"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    style={undefined}
  >
    <div
      aria-expanded={false}
      aria-haspopup="true"
      aria-owns="edit$Menu"
      className="rc-menu-submenu-title"
      onBlur={undefined}
      onClick={[Function]}
      onContextMenu={undefined}
      onFocus={undefined}
      onMouseDown={undefined}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onTouchStart={undefined}
      style={Object {}}
      title="Edit"
    >
      Edit
      <i
        className="rc-menu-submenu-arrow"
      />
    </div>
  </li>
  <li
    className="rc-menu-submenu rc-menu-submenu-horizontal"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    style={undefined}
  >
    <div
      aria-expanded={false}
      aria-haspopup="true"
      aria-owns="view$Menu"
      className="rc-menu-submenu-title"
      onBlur={undefined}
      onClick={[Function]}
      onContextMenu={undefined}
      onFocus={undefined}
      onMouseDown={undefined}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onTouchStart={undefined}
      style={Object {}}
      title="View"
    >
      View
      <i
        className="rc-menu-submenu-arrow"
      />
    </div>
  </li>
  <li
    className="rc-menu-submenu rc-menu-submenu-horizontal"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    style={undefined}
  >
    <div
      aria-expanded={false}
      aria-haspopup="true"
      aria-owns="cell$Menu"
      className="rc-menu-submenu-title"
      onBlur={undefined}
      onClick={[Function]}
      onContextMenu={undefined}
      onFocus={undefined}
      onMouseDown={undefined}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onTouchStart={undefined}
      style={Object {}}
      title="Cell"
    >
      Cell
      <i
        className="rc-menu-submenu-arrow"
      />
    </div>
  </li>
  <li
    className="rc-menu-submenu rc-menu-submenu-horizontal"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    style={undefined}
  >
    <div
      aria-expanded={false}
      aria-haspopup="true"
      aria-owns="runtime$Menu"
      className="rc-menu-submenu-title"
      onBlur={undefined}
      onClick={[Function]}
      onContextMenu={undefined}
      onFocus={undefined}
      onMouseDown={undefined}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onTouchStart={undefined}
      style={Object {}}
      title="Runtime"
    >
      Runtime
      <i
        className="rc-menu-submenu-arrow"
      />
    </div>
  </li>
  <li
    className="rc-menu-submenu rc-menu-submenu-horizontal"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    style={undefined}
  >
    <div
      aria-expanded={false}
      aria-haspopup="true"
      aria-owns="help$Menu"
      className="rc-menu-submenu-title"
      onBlur={undefined}
      onClick={[Function]}
      onContextMenu={undefined}
      onFocus={undefined}
      onMouseDown={undefined}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onTouchStart={undefined}
      style={Object {}}
      title="Help"
    >
      Help
      <i
        className="rc-menu-submenu-arrow"
      />
    </div>
  </li>
</ul>
`;
